### PR TITLE
Add context storage benchmarking

### PR DESCRIPTION
**Pull request by RLKRo**
# Description

Add `dff.utils.benchmark.context_storage` module which contains functions for benchmarking context storages.

This PR also includes all modules inside `dff.utils` (benchmarking + caching) during doc building.

# Checklist

- [x] I have covered the code with tests
- [x] I have added comments to my code to help others understand it
- [x] I have updated the documentation to reflect the changes
- [x] I have performed a self-review of the changes
- [x] Fix file inclusion in documentation (API ref references files located in `/utils/`, documentation will potentially fail if served at `/docs/build`)

### CODE REVIEW

File: dff/utils/benchmark/context_storage.py
Comment (pseusys) [inline dff/utils/benchmark/context_storage.py:63]: Maybe we should make message size random in length to reflect irregular nature of messages?

File: dff/utils/benchmark/context_storage.py
Comment (pseusys) [inline dff/utils/benchmark/context_storage.py:146]: This will never succeed if we read, say, 3 last requests from context storage only.
We should mind that OR manually set read/write policy of context storage to ALL.
But that should be done after merge of partial context updates only.

File: dff/utils/benchmark/context_storage.py
Comment (pseusys) [inline dff/utils/benchmark/context_storage.py:194]: Maybe output as string/string list instead of printing?

File: dff/utils/benchmark/context_storage.py
Comment (pseusys) [inline dff/utils/benchmark/context_storage.py:406]: What is actually the point? Something that is not an exception with a message can be thrown **and** catched here, like in C?

File: dff/utils/benchmark/context_storage.py
Comment (pseusys) [inline dff/utils/benchmark/context_storage.py:318]: Why does this method belongs th this class exactly? Why don't we make it a function?

File: dff/utils/benchmark/context_storage.py
Comment (pseusys) [inline dff/utils/benchmark/context_storage.py:511]: It seems to me (correct me if I'm wrong) that this file contains both *benchmarking* context storages and *retrieving* benchmark info (from files e.g.).
I consider these two to be completely different processes that might well be separated into two different source files.

File: dff/utils/turn_caching/singleton_turn_caching.py
Comment (pseusys) [inline dff/utils/turn_caching/singleton_turn_caching.py:5]: Seems to be unrelated to PR.

File: tests/utils/test_benchmark.py
Comment (pseusys) [inline tests/utils/test_benchmark.py:92]: Here and below: arguments can have a type.

File: tutorials/context_storages/8_db_benchmarking.py
Comment (pseusys) [inline tutorials/context_storages/8_db_benchmarking.py:136]: Benchmark is defined in a loop, so **logically** it should be in undefined state now.
Considering pythonic tricks here doesn't seem clear or appropriate to me.
Further more, I would say `benchmark` will be "YDB" at this point, not "Shelve".

File: utils/db_benchmark/benchmark_dbs.py
Comment (pseusys) [inline utils/db_benchmark/benchmark_dbs.py:23]: All these (and `YDB` on line 30) can and should be uncommented.

File: utils/db_benchmark/benchmark_new_format.py
Comment (pseusys) [inline utils/db_benchmark/benchmark_new_format.py:1]: Whole this file doesn't seem to be very appropriate to me: afaik this is the first release of benchmarking, so should we consider any previous development release and include format converting functions into production codebase?

File: utils/db_benchmark/benchmark_streamlit.py
Comment (pseusys) [inline utils/db_benchmark/benchmark_streamlit.py:10]: I think we should really add some descriptions on how to *use* these tabs (and the whole page) - onto the page itself.

**Review by pseusys — CHANGES_REQUESTED** [review]
I would like to specially request for one more important change:
Please, consider adding files created by both benchmarking and benchmark visualization website to `.gitignore` and `.dockerignore`.
Moreover, the project could benefit from creating separate `make` targets for clearing benchmark data (e.g. `make clear_benchmark`), at least this was a time-taking issue for me when I was testing context storages with benchmarks on a virtual machine.

There is also an issue I would like to address: what's the main purpose of benchmarks? Is it *testing* current context storages (in this case different paths to context storage sources can be removed) or is it *comparing* current context storages with something else? In latter case it is unclear what are we gonna compare them with: afaik we're not planning any structural changes to context storages. And, if we do, we might want to document how to create a comparison base backup of context storages (to compare the updated version with).
Please, also consider stopping and removing all the docker images after every benchmark: database structure can be different for different versions of context storages, and the name can be the same. This can be done by auto-modifying database paths for different versions of context storages or (and) by creating a special `make` target (e.g. `make docker_down`) that would also help us fighting disk space occupation by docker images and containers.

**Review by RLKRo — COMMENTED** [review]
> Please, consider adding files created by both benchmarking and benchmark visualization website to .gitignore and .dockerignore.

Added benchmark files to ignore files:
2420455d4b07b4e7c5a95f981c4b1334fa4cb877

> Moreover, the project could benefit from creating separate make targets for clearing benchmark data

Moved database directory inside the benchmark directory:
4263b7e21c3e8c9123b5ac3a9b738274b93bb405
So now everything generated by `benchmark_dbs.py` is stored in one place and can easily be removed.

> Please, also consider stopping and removing all the docker images after every benchmark

This seems out of scope for the task that benchmarking tries to solve.
Feels like these modifications should be made on the client side (by writing custom scripts for benchmarking).

File: tutorials/context_storages/8_db_benchmarking.py
Comment (RLKRo) [reply tutorials/context_storages/8_db_benchmarking.py:165]: Waiting for #141 to merge.

File: dff/utils/db_benchmark/benchmark.py
Comment (RLKRo) [reply dff/utils/db_benchmark/benchmark.py]: Added traceback to exception info in 8c419b5722cce710f7bb052958977554220c2e2e.

In streamlit it is displayed like this:
image

**Review by ruthenian8 — APPROVED** [review]

**Review by pseusys — APPROVED** [review]

**Review by kudep — APPROVED** [review]

**Pull request merged**
